README.md
# passport-ldapauth

[![Build Status](https://travis-ci.org/vesse/passport-ldapauth.svg)](https://travis-ci.org/vesse/passport-ldapauth)
[![npm](https://img.shields.io/npm/dm/passport-ldapauth.svg)](http://npmjs.com/package/passport-ldapauth)
[![Sponsored by Wakeone](https://img.shields.io/badge/sponsored%20by-wakeone-389fc1.svg)](https://wakeone.co)

[Passport](http://passportjs.org/) authentication strategy against LDAP / AD server. This module is a Passport strategy wrapper for [ldapauth-fork](https://github.com/vesse/node-ldapauth-fork).

This module lets you authenticate using LDAP or AD in your Node.js applications. By plugging into Passport, LDAP authentication can be integrated into any framework that supports Connect-style middleware.

## Install

```
npm install passport-ldapauth
```

## Usage
### Configure strategy

```javascript
var LdapStrategy = require('passport-ldapauth');

passport.use(new LdapStrategy({
    server: {
      url: 'ldap://localhost:389',
      ...
    }
  }));
```

* `server`: LDAP settings. These are passed directly to [ldapauth-fork](https://github.com/vesse/node-ldapauth-fork). See its documentation for all available options.
    * `url`: e.g. `ldap://localhost:389`
    * `bindDN`: e.g. `cn='root'`
    * `bindCredentials`: Password for bindDN
    * `searchBase`: e.g. `o=users,o=example.com`
    * `searchFilter`:  LDAP search filter, e.g. `(uid={{username}})`. Use literal `{{username}}` to have the given username used in the search.
    * `searchAttributes`: Optional array of attributes to fetch from LDAP server, e.g. `['displayName', 'mail']`. Defaults to `undefined`, i.e. fetch all attributes
    * `tlsOptions`: Optional object with options accepted by Node.js [tls](http://nodejs.org/api/tls.html#tls_tls_connect_options_callback) module.
* `usernameField`: Field name where the username is found, defaults to _username_
* `passwordField`: Field name where the password is found, defaults to _password_
* `credentialsLookup`: Optional, synchronous function that provides the login credentials from `req`. See [below](#credentialslookup) for more.
* `missingCredentialsStatus`: Returned HTTP status code when credentials could not be found in the request. Defaults to _400_
* `handleErrorsAsFailures`: When `true`, unknown errors and ldapjs emitted errors are handled as authentication failures instead of errors (default: `false`).
* `failureErrorCallback`: Optional, synchronous function that is called with the received error when `handleErrorsAsFailures` is enabled.
* `passReqToCallback`: When `true`, `req` is the first argument to the verify callback (default: `false`):

        passport.use(new LdapStrategy(..., function(req, user, done) {
            ...
            done(null, user);
          }
        ));

Note: you can pass a function instead of an object as `options`, see the [example below](#asynchronous-configuration-retrieval)

### Authenticate requests

Use `passport.authenticate()`, specifying the `'ldapauth'` strategy, to authenticate requests.

#### `authenticate()` options

In addition to [default authentication options](http://passportjs.org/guide/authenticate/) the following flash message options are available for `passport.authenticate()`:

 * `badRequestMessage`: missing username/password (default: 'Missing credentials')
 * `invalidCredentials`: `InvalidCredentialsError` and `/no such user/i` LDAP errors (default: 'Invalid username/password')
 * `noSuchObject`: `NoSuchObjectError` LDAP errors (default: 'Bad search base')
 * `userNotFound`: LDAP returns no error but also no user (default: 'Invalid username/password')
 * `constraintViolation`: user account is locked (default: 'Exceeded password retry limit, account locked')

And for [Microsoft AD messages](http://www-01.ibm.com/support/docview.wss?uid=swg21290631), these flash message options can also be used (used instead of `invalidCredentials` if matching error code is found):

 * `invalidLogonHours`: not being allowed to login at this current time (default: 'Not Permitted to login at this time')
 * `invalidWorkstation`: not being allowed to login from this current location (default: 'Not permited to logon at this workstation')
 * `passwordExpired`: expired password (default: 'Password expired')
 * `accountDisabled`: disabled account (default: 'Account disabled')
 * `accountExpired`: expired account (default: 'Account expired')
 * `passwordMustChange`: password change (default: 'User must reset password')
 * `accountLockedOut`: locked out account (default: 'User account locked')

## Express example

```javascript
var express      = require('express'),
    passport     = require('passport'),
    bodyParser   = require('body-parser'),
    LdapStrategy = require('passport-ldapauth');

var OPTS = {
  server: {
    url: 'ldap://localhost:389',
    bindDN: 'cn=root',
    bindCredentials: 'secret',
    searchBase: 'ou=passport-ldapauth',
    searchFilter: '(uid={{username}})'
  }
};

var app = express();

passport.use(new LdapStrategy(OPTS));

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({extended: false}));
app.use(passport.initialize());

app.post('/login', passport.authenticate('ldapauth', {session: false}), function(req, res) {
  res.send({status: 'ok'});
});

app.listen(8080);
```

### Active Directory over SSL example

Simple example config for connecting over `ldaps://` to a server requiring some internal CA certificate (often the case in corporations using Windows AD).

```javascript
var fs = require('fs');

var opts = {
  server: {
    url: 'ldaps://ad.corporate.com:636',
    bindDN: 'cn=non-person,ou=system,dc=corp,dc=corporate,dc=com',
    bindCredentials: 'secret',
    searchBase: 'dc=corp,dc=corporate,dc=com',
    searchFilter: '(&(objectcategory=person)(objectclass=user)(|(samaccountname={{username}})(mail={{username}})))',
    searchAttributes: ['displayName', 'mail'],
    tlsOptions: {
      ca: [
        fs.readFileSync('/path/to/root_ca_cert.crt')
      ]
    }
  }
};
...
```

## `credentialsLookup`

A synchronous function that receives the `req` object and returns an objec with keys `username` and `password` (or `name` and `pass`) can be provided. Note, that when this is provided the default lookup is not performed. This can be used to eg. enable basic auth header support:

```javascript
var basicAuth = require('basic-auth');
var ldapOpts = {
  server: { ... },
  credentialsLookup: basicAuth
}
```

## Asynchronous configuration retrieval

Instead of providing a static configuration object, you can pass a function as `options` that will take care of fetching the configuration. It will be called with the `req` object and a callback function having the standard `(err, result)` signature. Notice that the provided function will be called on every authenticate request.

```javascript
var getLDAPConfiguration = function(req, callback) {
  // Fetching things from database or whatever
  process.nextTick(function() {
    var opts = {
      server: {
        url: 'ldap://localhost:389',
        bindDN: 'cn=root',
        bindCredentials: 'secret',
        searchBase: 'ou=passport-ldapauth',
        searchFilter: '(uid={{username}})'
      }
    };

    callback(null, opts);
  });
};

var LdapStrategy = require('passport-ldapauth');

passport.use(new LdapStrategy(getLDAPConfiguration,
  function(user, done) {
    ...
    return done(null, user);
  }
));
```

## `ldapsearch`

[ldapsearch](http://linux.die.net/man/1/ldapsearch) is a great command line tool for testing your config. The user search query performed in the Express example above when user logging in has uid `john` is the same as the following `ldapsearch` call:

```bash
ldapsearch \
  -H ldap://localhost:389 \
  -x \
  -D cn=root \
  -w secret \
  -b ou=passport-ldapauth \
  "(uid=john)"
```

If the query does not return expected user the configuration is likely incorrect.

## License

MIT

`passport-ldapauth` has been partially sponsored by [Wakeone Ltd](https://wakeone.co/).


CHANGES.md
## Changes

* v3.0.0
    * Update ldapauth-fork to v5 which upgrade ldapjs to v2
* v2.1.4
    * Allow any version of @types/node
* v2.1.3
    * [#86](https://github.com/vesse/passport-ldapauth/pull/86) Allow configuring missing credentials response status.
* v2.1.2
    * [#80](https://github.com/vesse/passport-ldapauth/pull/80) Run error handler only once since the a new LdapAuth instance is created for every authenticate request.
* v2.1.1
    * Bump deps
* v2.1.0
    * [#77](https://github.com/vesse/passport-ldapauth/pull/77) Add `noSuchObject` error message
* v2.0.0
    * `ldapauth-fork` major version update now uses Bunyan logger
    * Added TypeScript type definitions
* v1.0.0
    * `ldapauth-fork` is now an event emitter. Emitted errors will cause authentication error.
    * [#38](https://github.com/vesse/passport-ldapauth/pull/38) Added option to handle erros as failures with `handleErrorsAsFailures`. Additionally a *synchronous* `failureErrorCallback` function that receives the error as argument can be provided.
* v0.6.0
    * Added option `credentialsLookup` that can be used eg. to add Basic Auth header parsing support.
* v0.5.0
    * Updated deps. ldapauth-fork update changes bind credentials handling to work better with falsy values needed in anonymous bind.
* v0.4.0
    * Updated ldapauth-fork which updates ldapjs to 1.0.0
* v0.3.1
    * [#35](https://github.com/vesse/passport-ldapauth/issues/35) - Show more specific error messages from Microsoft AD login errors if identified.
* v0.3.0
    * [#10](https://github.com/vesse/passport-ldapauth/issues/10) - Add support for fetching groups. While this is really coming from [ldapauth-fork](https://github.com/vesse/node-ldapauth-fork), updated the minor version of this library as well to draw attention to new features.
* v0.2.6
    * [#24](https://github.com/vesse/passport-ldapauth/pull/24) - Pass `req` to options function, enables request specific LDAP configuration.
* v0.2.5
    * [#21](https://github.com/vesse/passport-ldapauth/issues/21) - Handle `constraintViolationError` as a login failure instead of an error.
* v0.2.4
    * Inherit from [passport-strategy](https://github.com/jaredhanson/passport-strategy) like `passport-local` and others do.
* v0.2.3
    * Documentation using the same keys as ldapjs (bindDn and bindCredentials)
* v0.2.2
    * Allow configuring flash messages when calling `passport.authenticate()`
    * Return HTTP 400 when username or password is missing
* v0.2.1
    * Passport as peerDependency, prevents version incompatibility
* v0.2.0
    * [#8](https://github.com/vesse/passport-ldapauth/issues/8) - Possibility to provide a callback function instead of options object to constructor (contributed by Linagora)
    * Update Passport dependency to 0.2.0
    * Get rid of `var self = this;`
* v0.1.2
    * [#6](https://github.com/vesse/passport-ldapauth/issues/6) - Handle NoSuchObjectError as login failure.
* v0.1.1
    * Documentation changes due to renaming git repository of `ldapauth-fork`
* v0.1.0
    * Use [ldapauth-fork](https://github.com/vesse/node-ldapauth-fork) instead of
      [ldapauth](https://github.com/trentm/node-ldapauth)
        * ldapjs upgraded to 0.6.3
        * New options including `tlsOptions`
    * Refactored tests
* v0.0.6 (14 July 2013)
    * Fixes [#1](https://github.com/vesse/passport-ldapauth/issues/1)
    * Updated devDependencies
* v0.0.5 (16 April 2013)
    * Create LDAP client on every request to prevent socket being closed due
      to inactivity.
* v0.0.4 (14 April 2013)
    * Fixed passport-ldapauth version range.
* v0.0.3 (14 April 2013)
    * Initial release.


test/appserver.js
var express      = require('express'),
    passport     = require('passport'),
    LdapStrategy = require('passport-ldapauth').Strategy,
    bodyParser   = require('body-parser');

var server = null;

var init_passport = function(opts, testopts) {
  if (testopts.no_callback === true) {
    passport.use(new LdapStrategy(opts));
  } else {
    passport.use(new LdapStrategy(opts, function(user, cb) {
      return cb(null, user);
    }));
  }
};

passport.serializeUser(function(user, cb) {
  cb(null, user.dn);
});

passport.deserializeUser(function(dn, cb) {
  cb(null, {dn: dn});
});

exports.start = function(opts, testopts, cb) {

  var app = express();

  init_passport(opts, testopts);

  app.use(bodyParser.json());
  app.use(passport.initialize());

  app.post('/login', passport.authenticate('ldapauth', {session: false}), function(req, res) {
    res.send({status: 'ok'});
  });

  app.post('/custom-cb-login', function(req, res, next) {
    passport.authenticate('ldapauth', function(err, user, info) {
      if (err) return next(err);
      if (!user) return res.status(401).send(info);
      req.logIn(user, function(err) {
        if (err) return next(err);
        return res.json(user);
      })
    })(req, res, next);
  });

  if (typeof cb === 'function') return cb(app);
  return;
};

exports.close = function(cb) {
  if (server) server.close();
  server = null;
  if (typeof cb === 'function') return cb();
  return;
};


test/ldapserver.js
var ldap = require('ldapjs');

authorize = function(req, res, next) {
  return next();
};

var SUFFIX = 'ou=passport-ldapauth';
var server = null;

db = {
  'valid': {
    dn: 'cn=valid,ou=passport-ldapauth',
    attributes:  {
      uid:  'valid',
      name: 'Valid User'
    }
  }
};

exports.start = function(port, cb) {
  if (server) {
    if (typeof cb === 'function') return cb();
    return;
  }

  server = ldap.createServer();

  server.bind('cn=root', function(req, res, next) {
    if (req.dn.toString() !== 'cn=root' || req.credentials !== 'secret') {
      return next(new ldap.InvalidCredentialsError());
    }
    res.end();
    return next();
  });

  server.bind(SUFFIX, authorize, function(req, res, next) {
    var dn = req.dn.toString();
    if (dn !== 'cn=valid, ou=passport-ldapauth' || req.credentials !== 'valid') {
      return next(new ldap.InvalidCredentialsError());
    }
    res.end();
    return next();
  });

  server.search(SUFFIX, authorize, function(req, res, next) {
    if (req.filter.attribute === 'uid' && req.filter.value === 'valid') {
      res.send(db['valid']);
    } else if (req.filter.attribute === 'uid' && req.filter.value === 'ms-ad') {
      return next(new ldap.InvalidCredentialsError("0090308: LdapErr: DSID-0C09030B, comment: AcceptSecurityContext error, data 533, v893 HEX: 0x533 - account disabled"));
    } else if (req.filter.attribute === 'member' && req.filter.value === db.valid.dn) {
      res.send({
        dn: 'cn=Group 1, ou=passport-ldapauth',
        attributes: {
          name: 'Group 1'
        }
      });
      res.send({
        dn: 'cn=Group 2, ou=passport-ldapauth',
        attributes: {
          name: 'Group 2'
        }
      });
    }
    res.end();
    return next();
  });

  server.listen(port, function() {
    if (typeof cb === 'function') return cb();
  });
};

exports.close = function(cb) {
  if (server) server.close();
  server = null;
  if (typeof cb === 'function') return cb();
  return;
};

if (!module.parent) {
  exports.start(1389);
}


lib/passport-ldapauth/strategy.d.ts
// Type definitions for passport-ldapauth 2.0
// Project: https://github.com/vesse/passport-ldapauth
// Definitions by: Vesa Poikajärvi <https://github.com/vesse>
// TypeScript Version: 2.1

/// <reference types="node"/>

import { IncomingMessage } from 'http';
import { Options as LdapAuthOptions } from 'ldapauth-fork';
import {
    Strategy as PassportStrategy,
    AuthenticateOptions as PassportAuthenticateOptions
} from 'passport';

declare namespace Strategy {
    /**
     * Return value type for credentialsLookup
     */
    interface CredentialsLookupResult {
        username?: string;
        password?: string;
        user?: string;
        pass?: string;
    }

    /**
     * Credentials lookup function (eg. basic-auth)
     */
    type CredentialsLookup = (req: IncomingMessage) => CredentialsLookupResult;

    /**
     * Callback notified of an error when errors are handled as failures
     */
    type FailureErrorCallback = (err: any) => void;

    /**
     * passport-ldapauth options
     */
    interface Options {
        /**
         * ldapauth-fork connection options
         */
        server: LdapAuthOptions;
        /**
         * Form field name for username (default: username)
         */
        usernameField?: string;
        /**
         * Form field name for password (default: password)
         */
        passwordField?: string;
        /**
         * If set to true, request is passed to verify callback
         */
        passReqToCallback?: boolean;
        /**
         * Credentials lookup function to be used instead of default search from request
         */
        credentialsLookup?: CredentialsLookup;
        /**
         * Set to true to handle errors as login failures
         */
        handleErrorsAsFailures?: boolean;
        /**
         * Synchronous failure error callback for handling the failure if using handleErrorsAsFailures
         */
        failureErrorCallback?: FailureErrorCallback;
    }

    /**
     * Callback function returning the options if using OptionsFunction
     */
    type OptionsFunctionCallback = (error: any, options: Options) => void;

    /**
     * Callback for getting options dunamically for every authenticate call
     */
    type OptionsFunction = (req: IncomingMessage, callback: OptionsFunctionCallback) => void;

    /**
     * Flash message localizations for authenticate
     */
    interface AuthenticateOptions extends PassportAuthenticateOptions {
        badRequestMessage?: string;
        invalidCredentials?: string;
        userNotFound?: string;
        constraintViolation?: string;
        invalidLogonHours?: string;
        invalidWorkstation?: string;
        passwordExpired?: string;
        accountDisabled?: string;
        accountExpired?: string;
        passwordMustChange?: string;
        accountLockedOut?: string;
        noSuchObject?: string;
    }

    /**
     * Options that verify done callback can pass
     */
    interface VerifyOptions {
        message: string;
    }

    /**
     * Callback executed in the verify function once done
     *
     * @param error Possible error, setting this will result in error from Passport
     * @param user The user object, or false if there was no error but the authentication should be failed
     * @param options
     */
    type VerifyDoneCallback = (error: any, user?: any, options?: VerifyOptions) => void;

    /**
     * Verify callback when passReqToCallback = false
     */
    type VerifyCallback = (user: any, callback: VerifyDoneCallback) => void;

    /**
     * Verify callback when passReqToCallback = true
     */
    type VerifyCallbackWithRequest = (req: IncomingMessage, user: any, callback: VerifyDoneCallback) => void;
}

declare class Strategy implements PassportStrategy {
    /**
     * @param options Strategy options or function returning the options
     * @param verify User provided verify callback for checking the LDAP result
     */
    constructor(options: Strategy.Options | Strategy.OptionsFunction, verify?: Strategy.VerifyCallback | Strategy.VerifyCallbackWithRequest);

    /**
     * Name of the strategy
     */
    name: string;

    /**
     * @param req
     * @param options
     */
    authenticate(req: IncomingMessage, options?: Strategy.AuthenticateOptions): void;
}

export = Strategy;


lib/passport-ldapauth/strategy.js
'use strict';
var passport = require('passport-strategy');
var LdapAuth = require('ldapauth-fork');
var util     = require('util');

/**
 * Passport wrapper for ldapauth
 */


/**
 * Options callback callback (ie. the callback given if giving a callback
 * for options instead of an object)
 *
 * @callback optionsCallbackCallback
 * @param {(Error|undefined)} err - Possible error
 * @param {Object} options - Options object
 */
/**
 * Options callback
 *
 * @callback optionsCallback
 * @param {Object} req - HTTP request
 * @param {optionsCallbackCallback} callback - The callback returning the options
 */
/**
* Verify done callback
*
* @callback verifyDoneCallback
* @param {(Error|undefined)} err - Possible error
* @param {(Object|boolean)} user - The verified user or false if not allowed
* @param {Object} [info] info - Additional info message
*/
/**
* Found LDAP user verify callback
*
* @callback verifyCallback
* @param {Object} user - The user object from LDAP
* @param {verifyDoneCallback} callback - The verify callback
*/
/**
* Found LDAP user verify callback with request
*
* @callback verifyReqCallback
* @param {Object} req - The HTTP request
* @param {Object} user - The user object from LDAP
* @param {verifyDoneCallback} callback - The verify callback
*/
/**
 * @typedef credentialsLookupResult
 * @type {object}
 * @property {string} username - Username to use
 * @property {string} password - Password to use
 */
/**
 * @typedef credentialsLookupResultAlt
 * @type {object}
 * @property {string} user - Username to use
 * @property {string} pass - Password to use
 */
/**
 * Credentials lookup function
 *
 * @callback credentialsLookup
 * @param {Object} req - The HTTP request
 * @return {(credentialsLookupResult|credentialsLookupResultAlt)} - Found credentials
 */
/**
 * Synchronous function for doing something with an error if handling
 * errors as failures
 *
 * @callback failureErrorCallback
 * @param {Error} err - The error occurred
 */

/**
 * Add default values to options
 *
 * @private
 * @param {Object} options - Options object
 * @returns {Object} The given options with defaults filled
 */
var setDefaults = function(options) {
  options.usernameField || (options.usernameField = 'username');
  options.passwordField || (options.passwordField = 'password');
  return options;
};

/**
 * Strategy constructor
 * <br>
 *
 * The LDAP authentication strategy authenticates requests based on the
 * credentials submitted through an HTML-based login form.
 * <br>
 *
 * Applications may supply a `verify` callback which accepts `user` object
 * and then calls the `done` callback supplying a `user`, which should be set
 * to `false` if user is not allowed to authenticate. If an exception occured,
 * `err` should be set.
 * <br>
 *
 * Options can be also given as function that accepts a callback end calls it
 * with error and options arguments. Notice that the callback is executed on
 * every authenticate call.
 * <br>
 *
 * Example:
 * <pre>
 * var LdapStrategy = require('passport-ldapauth').Strategy;
 * passport.use(new LdapStrategy({
 *     server: {
 *       url: 'ldap://localhost:389',
 *       bindDN: 'cn=root',
 *       bindCredentials: 'secret',
 *       searchBase: 'ou=passport-ldapauth',
 *       searchFilter: '(uid={{username}})',
 *       reconnect: true
 *     }
 *   },
 *   function(user, done) {
 *     return cb(null, user);
 *   }
 * ));
 * </pre>
 * @constructor
 * @param {(Object|optionsCallback)} options - Configuration options or options returning function
 * @param {Object} options.server - [ldapauth-fork options]{@link https://github.com/vesse/node-ldapauth-fork}
 * @param {string} [options.usernameField=username] - Form field name for username
 * @param {string} [options.passwordField=password] - Form field name for password
 * @param {Number} [options.missingCredentialsStatus=400] - HTTP status code returned when credentials are missing from the request
 * @param {boolean} [options.passReqToCallback=false] - If true, request is passed to verify callback
 * @param {credentialsLookup} [options.credentialsLookup] - Credentials lookup function to use instead of default
 * @param {boolean} [options.handleErrorAsFailures=false] - Set to true to handle errors as login failures
 * @param {failureErrorCallback} [options.failureErrorCallback] - Function receives the occurred error when errors handled as failures
 * @param {(verifyCallback|verifyReqCallback|undefined)} [verify] - User verify callback
 */
var Strategy = function(options, verify) {
  // We now accept function as options as well so we cannot assume anymore
  // that a call with a function parameter only would have skipped options
  // and just provided a verify callback
  if (!options) {
    throw new Error('LDAP authentication strategy requires options');
  }

  this.options    = null;
  this.getOptions = null;

  if (typeof options === 'object') {
    this.options = setDefaults(options);
    this.ldap = new LdapAuth(this.options.server);
  } else if (typeof options === 'function') {
    this.getOptions = options;
  }

  passport.Strategy.call(this);

  this.name   = 'ldapauth';
  this.verify = verify;
};

util.inherits(Strategy, passport.Strategy);

/* eslint-disable */
/**
 * Get value for given field from given object. Taken from passport-local,
 * copyright 2011-2013 Jared Hanson
 *
 * @private
 * @param {Object} obj - The HTTP request object
 * @param {string} field - The field name to look for
 * @returns {string|null} - Found value for the field or null
 */
var lookup = function(obj, field) {
  var i, len, chain, prop;
  if (!obj) { return null; }
  chain = field.split(']').join('').split('[');
  for (i = 0, len = chain.length; i < len; i++) {
    prop = obj[chain[i]];
    if (typeof(prop) === 'undefined') { return null; }
    if (typeof(prop) !== 'object') { return prop; }
    obj = prop;
  }
  return null;
};
/* eslint-enable */

/**
 * Verify the outcome of caller verify function - even if authentication (and
 * usually authorization) is taken care by LDAP there may be reasons why
 * a verify callback is provided, and again reasons why it may reject login
 * for a valid user.
 *
 * @private
 * @returns {undefined}
 */
var verify = function() {
  // Callback given to user given verify function.
  return function(err, user, info) {
    if (err) {
      return this.error(err);
    }
    if (!user) {
      return this.fail(info);
    }
    return this.success(user, info);
  }.bind(this);
};

/**
 * Execute failureErrorCallback if provided
 *
 * @private
 * @param {Error} err - The error occurred
 * @returns {undefined}
 */
var handleErrorAsFailure = function(err) {
  if (typeof this.options.failureErrorCallback === 'function') {
    this.options.failureErrorCallback(err);
  }
  return this.fail(err, 500);
};

/**
 * The actual authenticate implementation
 *
 * @private
 * @param {Object} req - The HTTP request
 * @param {Object} [options] - Flash messages
 * @returns {undefined}
 */
var handleAuthentication = function(req, options) {
  var username;
  var password;
  var ldap;
  var errorHandler;
  options || (options = {}); // eslint-disable-line no-param-reassign

  if (typeof this.options.credentialsLookup === 'function') {
    var credentials = this.options.credentialsLookup(req);
    if (credentials != null) {
      // name and pass as a courtesy for those who use basic-auth directly as
      // they're likely the main user group.
      username = credentials.username || credentials.name;
      password = credentials.password || credentials.pass;
    }
  } else {
    username = lookup(req.body, this.options.usernameField) || lookup(req.query, this.options.usernameField);
    password = lookup(req.body, this.options.passwordField) || lookup(req.query, this.options.passwordField);
  }

  if (!username || !password) {
    return this.fail({ message: options.badRequestMessage || 'Missing credentials' }, this.options.missingCredentialsStatus || 400);
  }

  errorHandler = this.options.handleErrorsAsFailures === true ? handleErrorAsFailure.bind(this) : this.error.bind(this);

  /**
   * AD possible messages
   * http://www-01.ibm.com/support/docview.wss?uid=swg21290631
   */
  var messages = {
    '530': options.invalidLogonHours || 'Not Permitted to login at this time',
    '531': options.invalidWorkstation || 'Not permited to logon at this workstation',
    '532': options.passwordExpired || 'Password expired',
    '533': options.accountDisabled || 'Account disabled',
    '534': options.accountDisabled || 'Account disabled',
    '701': options.accountExpired || 'Account expired',
    '773': options.passwordMustChange || 'User must reset password',
    '775': options.accountLockedOut || 'User account locked',
    default: options.invalidCredentials || 'Invalid username/password'
  };

  ldap = this.ldap ? this.ldap : new LdapAuth(this.options.server);
  ldap.once('error', errorHandler);
  ldap.on('error', function() { /* Ignored */ });
  ldap.authenticate(username, password, function(err, user) {
    ldap.close(function(){
      // We don't care about the closing
    });

    if (err) {
      // Invalid credentials / user not found are not errors but login failures
      if (err.name === 'InvalidCredentialsError' || err.name === 'NoSuchObjectError' || (typeof err === 'string' && err.match(/no such user/i))) {
        var message;
        if (err.name === 'NoSuchObjectError') {
          message = options.noSuchObject || 'Bad search base';
        } else {
          message = options.invalidCredentials || 'Invalid username/password';
        }

        if (err.message) {
          var ldapComment = err.message.match(/data ([0-9a-fA-F]*), v[0-9a-fA-F]*/);
          if (ldapComment && ldapComment[1]){
            message = messages[ldapComment[1]] || messages['default'];
          }
        }
        return this.fail({ message: message }, 401);
      }
      if (err.name === 'ConstraintViolationError'){
        return this.fail({ message: options.constraintViolation || 'Exceeded password retry limit, account locked' }, 401);
      }

      // Other errors are (most likely) real errors
      return errorHandler(err);
    }

    if (!user) {
      return this.fail({ message: options.userNotFound || 'Invalid username/password' }, 401);
    }

    // Execute given verify function
    if (this.verify) {
      if (this.options.passReqToCallback) {
        return this.verify(req, user, verify.call(this));
      } else {
        return this.verify(user, verify.call(this));
      }
    } else {
      return this.success(user);
    }
  }.bind(this));
};

/**
 * Authenticate the request coming from a form or such.
 *
 * @param {Object} req - The HTTP request
 * @param {Object} [options] - Authentication options (flash messages). All messages have default values.
 * @param {string} [options.badRequestMessage] - Message for missing username/password
 * @param {string} [options.invalidCredentials] - Message for InvalidCredentialsError, NoSuchObjectError, and /no such user/ LDAP errors
 * @param {string} [options.userNotFound] - Message for user not found
 * @param {string} [options.constraintViolation] - Message when account is locked (or other constraint violation)
 * @param {string} [options.invalidLogonHours] - Message for Windows AD invalidLogonHours error
 * @param {string} [options.invalidWorkstation] - Message for Windows AD invalidWorkstation error
 * @param {string} [options.passwordExpired] - Message for Windows AD passwordExpired error
 * @param {string} [options.accountDisabled] - Message for Windows AD accountDisabled error
 * @param {string} [options.accountExpired] - Message for Windows AD accountExpired error
 * @param {string} [options.passwordMustChange] - Message for Windows AD passwordMustChange error
 * @param {string} [options.accountLockedOut] - Message for Windows AD accountLockedOut error
 * @returns {undefined}
 */
Strategy.prototype.authenticate = function(req, options) {
  if ((typeof this.options === 'object') && (!this.getOptions)) {
    return handleAuthentication.call(this, req, options);
  }

  var callback = function(err, configuration) {
    if (err) {
      return this.fail(err);
    }

    this.options = setDefaults(configuration);
    handleAuthentication.call(this, req, options);
  };

  // Added functionality: getOptions can accept now up to 2 parameters
  if (this.getOptions.length === 1) { // Accepts 1 parameter, backwards compatibility
    this.getOptions(callback.bind(this));
  } else { // Accepts 2 parameters, pass request as well
    this.getOptions(req, callback.bind(this));
  }
};

module.exports = Strategy;
